controllers/integration/integration_adapter.go

```diff
@@ -0,0 +1,310 @@
+func (a *Adapter) findMatchingApplicationSnapshot() (*releasev1alpha1.ApplicationSnapshot, error) {
+	var allApplicationSnapshots *[]releasev1alpha1.ApplicationSnapshot
+	expectedApplicationSnapshot, err := a.createApplicationSnapshotForPipelineRun()
+	if err != nil {
+		return nil, err
+	}
+	allApplicationSnapshots, err = a.getAllApplicationSnapshots()
+	if err != nil {
+		return nil, err
+	}
+	for _, foundApplicationSnapshot := range *allApplicationSnapshots {
+		if a.compareApplicationSnapshots(expectedApplicationSnapshot, &foundApplicationSnapshot) {
+			return &foundApplicationSnapshot, nil
+		}
+	}
+	return nil, nil
+}
+
+// getAllApplicationComponents loads from the cluster all Components associated with the given Application.
+// If the Application doesn't have any Components or this is not found in the cluster, an error will be returned.
+func (a *Adapter) getAllApplicationComponents(application *hasv1alpha1.Application) (*[]hasv1alpha1.Component, error) {
+	applicationComponents := &hasv1alpha1.ComponentList{}
+	opts := []client.ListOption{
+		client.InNamespace(application.Namespace),
+		client.MatchingFields{"spec.application": application.Name},
+	}
+
+	err := a.client.List(a.context, applicationComponents, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &applicationComponents.Items, nil
+}
+
+// getAllApplicationReleaseLinks returns the ReleaseLinks used by the application being processed. If matching
+// ReleaseLinks are not found, an error will be returned.
+func (a *Adapter) getAllApplicationReleaseLinks() (*[]releasev1alpha1.ReleaseLink, error) {
+	releaseLinks := &releasev1alpha1.ReleaseLinkList{}
+	opts := []client.ListOption{
+		client.InNamespace(a.application.Namespace),
+		client.MatchingFields{"spec.application": a.application.Name},
+	}
+
+	err := a.client.List(a.context, releaseLinks, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &releaseLinks.Items, nil
+}
+
+// getApplicationSnapshot returns the all ApplicationSnapshots in the Application's namespace nil if it's not found.
+// In the case the List operation fails, an error will be returned.
+func (a *Adapter) getAllApplicationSnapshotReleases(applicationSnapshot *releasev1alpha1.ApplicationSnapshot) (*[]releasev1alpha1.Release, error) {
+	releases := &releasev1alpha1.ReleaseList{}
+	opts := []client.ListOption{
+		client.InNamespace(applicationSnapshot.Namespace),
+		client.MatchingFields{"spec.applicationSnapshot": applicationSnapshot.Name},
+	}
+
+	err := a.client.List(a.context, releases, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &releases.Items, nil
+}
+
+// createApplicationSnapshotForPipelineRun creates the ApplicationSnapshot for a given PipelineRun
+func (a *Adapter) createApplicationSnapshotForPipelineRun() (*releasev1alpha1.ApplicationSnapshot, error) {
+	applicationComponents, err := a.getAllApplicationComponents(a.application)
```

Comment:
In general, I'd change the creation methods to receive everything they need (in this case it should receive the components as a parameter) and handle the creation in the cluster within the function.